### PR TITLE
GUI: Qt port

### PR DESCRIPTION
**Pull request by fdde**

### CODE REVIEW

**Comment by farmboy0** [comment]
Hi,
thanks for the PR.
Could you please follow the projects code style, its documented at https://wiki.xoreos.org/index.php?title=Code_Formatting_Conventions
It would also make this PR easier to review if you squash commits that only fix up earlier ones.

**Comment by farmboy0** [comment]
2 Things I have already noticed:
1. There is a menu item MainWindow in the File menu, is this supposed to be Quit?
2. Adding a file, removing it, adding a second file will make the first file show up in the tree with the second one.

**Comment by farmboy0** [comment]
Please dont ignore my first comment. It is by far the more important one.

**Comment by fdde** [comment]
Sorry, I hadn't been checking my notifications. I'll get started on these.

**Comment by fdde** [comment]
So I coincidentally fixed the 2 things you mentioned at around the same time.

I've added a commit that fixes the style based on running the .uncrustifyrc, some regexes, adding comments, and reordering headers.

Regarding squashing the commits: it seems that most of the commits that fix up earlier ones aren't consecutive and I'm not good with git.

**Comment by farmboy0** [comment]
Thanks. I'll take a look asap.

**Comment by DrMcCoy** [comment]
The format you use for commit messages isn't quite correct: you need to have an empty line between the first line and the rest.

I.e. instead of

``` 
GUI: Remove QString functions from util and ustring
This means we have to convert to std::string/c_str but I think it's better this way.
Also don't set the _rootPath in the main window constructor, because it's set in setTreeViewModel anyway.
```

It needs to be

``` 
GUI: Remove QString functions from util and ustring

This means we have to convert to std::string/c_str but I think it's better this way.
Also don't set the _rootPath in the main window constructor, because it's set in setTreeViewModel anyway.
```

**Comment by DrMcCoy** [comment]
Ah, okay, seems like Verdigris uses a Qt 5.5 feature...

**Comment by fdde** [comment]
Fixed commit messages.

**Comment by DrMcCoy** [comment]
Okay, I managed to set up Travis CI with an updated gcc and Qt.

It's in this branch: https://github.com/xoreos/phaethon/tree/qttravis , which is mostly your commits. https://github.com/xoreos/phaethon/commit/3b803a2f511c296a3808f92e72dde95036f37ea0 is the commit in question.

Likewise, https://github.com/xoreos/phaethon/commit/9a1188384d1ee81d4deddc7e39315c208df7b083 is necessary to make our CMake build system happy and https://github.com/xoreos/phaethon/commit/8107b0f00e2e45aaff4b537a602f4962817de9ce is needed for clang (because, unlike gcc, clang still considers C++03 the default).

 If you could pull those into your branch/PR, that would be great. The other three commits for boost.m4 aren't strictly necessary.

**Comment by DrMcCoy** [comment]
I'm still not 100% happy with your commit messages, though.

- https://github.com/xoreos/phaethon/pull/5/commits/da49c8fe4935e29aa4b71a983f944eb901114ce3 "GUI Updates" is just too unspecific. You might also want to split those three bullet points into their own commits
- The same with https://github.com/xoreos/phaethon/pull/5/commits/fb603a5f6098ad6eb463eab8b6eefe301348699a "GUI: Fix some fixme's"
- https://github.com/xoreos/phaethon/pull/5/commits/3a88238a9e7b4cb7ed92b0bd004ebde3d5e027ac is even completely missing a short description
- Several first lines are still too long. You can see that by them being cut off by GitHub. Namely, https://github.com/xoreos/phaethon/pull/5/commits/11acf6a9973e46d47a4b8ba1fda11178ae26d9dc, https://github.com/xoreos/phaethon/pull/5/commits/909bf72b5a2b07d2d62b9b4111844242080b26f1, https://github.com/xoreos/phaethon/pull/5/commits/06cc99e482cde3ac9b08b8cffdace3efa030ceac and https://github.com/xoreos/phaethon/pull/5/commits/e2252fc7fc8623987bd2e36f53b9ae49ef7525b7

A lot of your "fixes" commits should just be merged into the commit that adds the code to begin with. But no worries, if you're not able to do that, it's fine. I can do that, once we're otherwise happy :)

File: src/gui/resourcetreeitem.h [hunk 1]
```diff
@@ -6,76 +6,88 @@
...
     Images::Decoder *getImage();
     Images::Decoder *getImage(Common::SeekableReadStream &res, Aurora::FileType type);
-    Common::SeekableReadStream *getResourceData() const;
-    Data &getData();
-    uint64 getSoundDuration() const;
     Sound::AudioStream *getAudioStream() const;
+    uint64 getSoundDuration() const;
+
+private:
+    /** Model information. **/
```
Comment (DrMcCoy) [inline src/gui/resourcetreeitem.h]: `/** **/` denotes a [Doxygen](https://xoreos.org/doxygen/)-comment. Those don't span multiple items, only the one immediately following the comment. In this case, the `_parent` member variable.

Here, you probably want to use normal comments instead.

File: src/gui/resourcetreeitem.h [hunk 2]
```diff
@@ -56,24 +56,36 @@ struct Archive {
...
-    Aurora::FileType _fileType;
-    Aurora::ResourceType _resourceType;
     qint64 _size;
+};
+
+class ResourceTreeItem {
+private:
+    ResourceTreeItem *_parent;
+    QList<ResourceTreeItem*> _children;
+    QString _name; // The filename. It's what the tree view displays.
```
Comment (DrMcCoy) [inline src/gui/resourcetreeitem.h]: This comment here, on the other hand, looks like it should be a Doxygen comment. Use `///<` to create a Doxygen comment that points to the preceeding item on the same line.

Like so:
```
QString _name; ///< The filename. It's what the tree view displays.
```

**Comment by DrMcCoy** [comment]
Also, your indenting style is completely wrong. You need to use tabs, not spaces, to indent.

File: src/gui/mainwindow.h [hunk 3]
```diff
@@ -25,98 +25,119 @@
...
-	wxDECLARE_EVENT_TABLE();
+	// resource preview
+	Common::ScopedPtr<PanelPreviewEmpty> _panelPreviewEmpty;
+	Common::ScopedPtr<PanelPreviewImage> _panelPreviewImage;
+	Common::ScopedPtr<PanelPreviewSound> _panelPreviewSound;
+	Common::ScopedPtr<PanelResourceInfo> _panelResourceInfo;
 };
 
-} // End of namespace GUI
+} // namespace GUI
```
Comment (farmboy0) [inline src/gui/mainwindow.h]: Please revert this change.

File: src/gui/resourcetree.h [hunk 4]
```diff
@@ -22,16 +22,26 @@
  *  Phaethon's tree of game resource files.
  */
 
-#ifndef RESOURCETREE_H
-#define RESOURCETREE_H
+#ifndef TREEMODEL_H
```
Comment (farmboy0) [inline src/gui/resourcetree.h]: Please revert this change.

File: src/gui/resourcetree.h [hunk 5]
```diff
@@ -45,128 +55,91 @@ namespace Images {
...
+	int rowCount(const QModelIndex &parent = QModelIndex()) const override;
 
-	wxDECLARE_EVENT_TABLE();
+	/** Add children to the given index.
+	*/
+	void fetchMore(const QModelIndex &index);
 };
 
-} // End of namespace GUI
+} // namespace GUI
```
Comment (farmboy0) [inline src/gui/resourcetree.h]: Please revert this change.

File: src/gui/resourcetree.h [hunk 6]
```diff
@@ -45,128 +55,91 @@ namespace Images {
...
 
-#endif // RESOURCETREE_H
+#endif // TREEMODEL_H
```
Comment (farmboy0) [inline src/gui/resourcetree.h]: Please revert this change.

File: src/phaethon.cpp [hunk 7]
```diff
@@ -22,30 +22,81 @@
...
+
+	int argc = 1; // QApplication requires it to be at least 1
+    char empty[] = ""; // silence -Wwrite-string warning
+    char *argv[] = {empty}; // must be at least 1
+    QApplication app(argc, argv);
+
+    GUI::MainWindow mainWindow(0, Version::getProjectNameVersion(), QSize(800, 600), path.c_str());
+    mainWindow.show();
+
+    app.exec();
```
Comment (farmboy0) [inline src/phaethon.cpp]: TABs for indenting, spaces for aligning.

**Comment by fdde** [comment]
Updated as per farmboy0's feedback.

**Comment by DrMcCoy** [comment]
Hey, just wanted to check in with you. I sent you an email on April 19. Might be that gmail put it into the spam folder, so if you haven't seen it, could you please check? Cheers!
